### ISSUE TITLE
New/updated lane types (further LaneType refactoring)

### ISSUE THREAD

**Issue by j-d-b** [usdot-jpo-ode/wzdx#149]
Lanes and lane types have been a topic of discussion in the past three WZDx development cycles. Much of the effort was focused on cleaning up the existing lane types—there is still work to be done in this realm.

In addition to cleanup, v3.0 saw the inclusion of three new lane types, `hov-lane`, `center-left-turn-lane`, and `reversible lane`. These were added to enable functionality that was present in previous versions.

In recent subgroup discussions, many members proposed ideas for more lane types; this issue is meant to be the place to discuss those.

Some examples for consideration below:

- `exit-only-lane`
- `right-turn-only-lane`
- `left-turn-only-lane`

**Comment by semosher** [comment]
For reference, these are the enumerations TxDOT uses for it's lanes on highways.

- MainLane
- EntranceRamp
- ExitRamp
- EntranceAndExitRamp
- ExitAndEntranceRamp
- Shoulder
- HOV
- Turnaround
- ExpressLane
- LeftInterchangeRamp
- RightInterchangeRamp
- InterchangeEntranceRamp
- HovMainLane
- HovEntranceRamp
- HovExitRamp
- HovEntranceAndExitRamp
- HovExitAndEntranceRamp
- HovLeftInterchangeRamp
- HovRightInterchangeRamp
- HovInterchangeEntranceRamp

**Comment by j-d-b** [comment]
@semosher is "Ramp" used generally, as in, it doesn't have to be a gain in elevation?

Maybe `exit-ramp` and `entrance-ramp` are better that `exit-lane` and `entrance-lane` (includes "ramps") proposed in #147.

**Comment by semosher** [comment]
@j-d-b "Ramp" is used generally without the need for a change in elevation.

**Comment by j-d-b** [comment]
@semosher that makes sense, but is different from what @davidcraig4000 preferred, which is what #147 was based on. I am curious if others have thoughts. TMDD has both `right/left-exit-lane` and `right/left-exit-ramp`

**Comment by vchandranNTAM** [comment]
As long as we have the standards across the board, we should use these lane-types. The problem is enumerated types have the risk of having that one value for a DOT that does not belong (due to their own definitions) and that will break the bank and adoption of the system will suffer. 

I agree with @j-d-b - we should go with a standard if it exists (or) we should stay at a higher level that can accommodate every potential value as a "sub" value. 

The additional question to be asked here is who would be interested in this data point? I can see some decisions being made based on the lane type.

**Comment by j-d-b** [comment]
The following is copied from #147 for reference here:

---

Here is an incomplete proposal for even simpler lane types, for discussion:

```
General/Normal/Travel/Main/Vehicle (need to choose one term)
Sidewalk
Bike lane
Exit ramp/lane
Entrance ramp/lane
Shoulder
Parking (new functionality)
Median (new functionality)
Center left turn lane
```

Compared to the list in the above comment, this list drops the ability to tell the direction of egress/ingress of an exit/entrance ramp/lane from the lane type (maybe this doesn't matter). It also drops the turning lanes, which are incomplete and not too helpful anyway as there is a lot of functionality missing ("left turn only",  "straight only", "right or left turns" etc.). As done in SAE J2735, perhaps the turns (what maneuvers are allowed) should not be part of the lane type and maybe we don't need to address that level of detail in WZDx.

**Comment by j-d-b** [comment]
What remains for discussion before I do the implementation and PR is:

1. Is there any opposition to dropping the `left` and `right` from exit and entrance?
2. Is there any opposition to dropping "exit lane" and just using the "exit ramp" terminology?
3. For general, normal lanes, which lane type name is preferred (this will replace/rename `lane`)? Options:
        1. General
        2. Normal
        3. Travel
        4. Main
        5. Vehicle

**Comment by lynnerandolph** [comment]
Not that I'm advocating for using this, necessarily, but wanted to provide the TMDD v3 LaneRoadway type. My feeling is that TMDD v3 decided to be super specific about everything and is very complicated. But, since it's a national standard, throwing it in the mix. It also unions with a "local" version which means it can contain even more than this. I'm going to leave another, separate comment.

	<xs:simpleType name="LaneRoadway">
		<xs:union>
			<xs:simpleType>
				<xs:restriction base="xs:unsignedInt">
					<xs:minInclusive value="8192"/>
					<xs:maxInclusive value="8447"/>
				</xs:restriction>
			</xs:simpleType>
			<xs:simpleType>
				<xs:restriction base="xs:string">
					<xs:enumeration value="all roadways" id="_8193"/>
					<xs:enumeration value="through lanes" id="_8194"/>
					<xs:enumeration value="left lane" id="_8195"/>
					<xs:enumeration value="right lane" id="_8196"/>
					<xs:enumeration value="center lane" id="_8197"/>
					<xs:enumeration value="middle lanes" id="_8198"/>
					<xs:enumeration value="middle two lanes" id="_8199"/>
					<xs:enumeration value="right turning lanes" id="_8200"/>
					<xs:enumeration value="left turning lanes" id="_8201"/>
					<xs:enumeration value="upper deck lanes" id="_8236"/>
					<xs:enumeration value="lower deck lanes" id="_8237"/>
					<xs:enumeration value="reversible lanes" id="_8238"/>
					<xs:enumeration value="right exit lanes" id="_8239"/>
					<xs:enumeration value="left exit lanes" id="_8240"/>
					<xs:enumeration value="right merging lanes" id="_8241"/>
					<xs:enumeration value="left merging lanes" id="_8242"/>
					<xs:enumeration value="right exit ramp" id="_8202"/>
					<xs:enumeration value="right second exit ramp" id="_8243"/>
					<xs:enumeration value="right entrance ramp" id="_8203"/>
					<xs:enumeration value="right second entrance ramp" id="_8245"/>
					<xs:enumeration value="left exit ramp" id="_8204"/>
					<xs:enumeration value="left second exit ramp" id="_8244"/>
					<xs:enumeration value="left entrance ramp" id="_8205"/>
					<xs:enumeration value="left second entrance ramp" id="_8246"/>
					<xs:enumeration value="escape ramp" id="_8234"/>
					<xs:enumeration value="hard shoulder" id="_8206"/>
					<xs:enumeration value="soft shoulder" id="_8207"/>
					<xs:enumeration value="right shoulder" id="_8208"/>
					<xs:enumeration value="left shoulder" id="_8209"/>
					<xs:enumeration value="highways" id="_8235"/>
					<xs:enumeration value="right hand parallel lanes" id="_8210"/>
					<xs:enumeration value="left hand parallel lanes" id="_8211"/>
					<xs:enumeration value="connecting lanes" id="_8212"/>
					<xs:enumeration value="express lanes" id="_8213"/>
					<xs:enumeration value="local lanes" id="_8214"/>
					<xs:enumeration value="toll lanes" id="_8215"/>
					<xs:enumeration value="electronic toll lanes" id="_8216"/>
					<xs:enumeration value="toll plaza" id="_8217"/>
					<xs:enumeration value="inspection lane" id="_8218"/>
					<xs:enumeration value="hOV lanes" id="_8219"/>
					<xs:enumeration value="bus lanes" id="_8220"/>
					<xs:enumeration value="carpool lanes" id="_8221"/>
					<xs:enumeration value="truck lanes" id="_8222"/>
					<xs:enumeration value="emergency lanes" id="_8223"/>
					<xs:enumeration value="passing lanes" id="_8224"/>
					<xs:enumeration value="climbing lanes" id="_8225"/>
					<xs:enumeration value="slow lane" id="_8226"/>
					<xs:enumeration value="service road" id="_8227"/>
					<xs:enumeration value="cycle lane" id="_8228"/>
					<xs:enumeration value="bridge" id="_8229"/>
					<xs:enumeration value="overpass" id="_8230"/>
					<xs:enumeration value="elevated lanes" id="_8231"/>
					<xs:enumeration value="underpass" id="_8232"/>
					<xs:enumeration value="tunnel" id="_8233"/>
				</xs:restriction>
			</xs:simpleType>
			<xs:simpleType>
				<xs:restriction base="xs:string">
					<xs:pattern value="\[.+\].*"/>
				</xs:restriction>
			</xs:simpleType>
			<xs:simpleType>
				<xs:restriction base="local:LaneRoadway"/>
			</xs:simpleType>
		</xs:union>

**Comment by lynnerandolph** [comment]
1. I have no objection to dropping the left/right as there is already ordering to the lanes.
2. I'm fine with exit ramp
3. To me, main has a specific meaning which isn't inclusive of special travel lanes. I vote for "Travel".
> 
>     1. Is there any opposition to dropping the `left` and `right` from exit and entrance?
> 
>     2. Is there any opposition to dropping "exit lane" and just using the "exit ramp" terminology?
> 
>     3. For general, normal lanes, which lane type name is preferred (this will replace/rename `lane`)? Options:
>        1. General
>        2. Normal
>        3. Travel
>        4. Main
>        5. Vehicle

**Comment by j-d-b** [comment]
@lynnerandolph WZDx was initially based on TMDD, [see for reference the lane type enumerated type from WZDx v1.1](https://github.com/usdot-jpo-ode/wzdx/blob/v1.1/data-tables/enumerated-types.md).

WZDx LaneType has moved away from TMDD the past few releases, most notably after WZDx added the `Lane` object which allowed describing the lanes with more detail than just the lane type. TMDD's LaneRoadway is used in a context where the value itself is used to determine the lane's location on the roadway whereas WZDx has `order`. Also values like `hov lanes` are covered by WZDx as restrictions, which matches SAE J2735. Thus the use case of LaneRoadway doesn't match closely enough with WZDx's LaneType.

**Comment by j-d-b** [comment]
> 1. I have no objection to dropping the left/right as there is already ordering to the lanes.
> > 1. Is there any opposition to dropping the `left` and `right` from exit and entrance?

@lynnerandolph regarding the above, the `left` and `right` that remain do not refer to the lane's position (determined by `order`). These directions refer to the direction of ingress/egress, see the current [LaneType](https://github.com/usdot-jpo-ode/wzdx/blob/v3.1/spec-content/enumerated-types/LaneType.md) description for `left-exit-ramp` for example:

> An exit ramp with an egress on the left in the direction of flow at an interchange.

**Comment by yc-uw** [comment]
> 
> 
> What remains for discussion before I do the implementation and PR is:
> 
>     1. Is there any opposition to dropping the `left` and `right` from exit and entrance?
> 
>     2. Is there any opposition to dropping "exit lane" and just using the "exit ramp" terminology?
> 
>     3. For general, normal lanes, which lane type name is preferred (this will replace/rename `lane`)? Options:
>        1. General
>        2. Normal
>        3. Travel
>        4. Main
>        5. Vehicle

Wisconsin is reviewing this and will provide some feed backs in a couple of weeks.

**Comment by davidcraig4300** [comment]
I like General and Normal.  The others (Travel, Main, Vehicle) don't really make sense to me.  I believe our internal map terminology uses Normal, but i believe General would also work well.

**Comment by j-d-b** [comment]
Implemented in #188.
